samza-api/src/main/java/org/apache/samza/operators/Windows.java

```diff
@@ -0,0 +1,202 @@
+
+    /**
+     * function to calculate the output value from the input message and the current output value
+     */
+    private final BiFunction<M, WV, WV> aggregator;
+
+    /**
+     * trigger condition that determines when to send out the output value in a {@link WindowOutput} message
+     */
+    private Trigger<M, WindowState<WV>> trigger = null;
+
+    //TODO: need to create a set of {@link StoreFunctions} that is default to input {@link Message} type for {@link Window}
+    private Operators.StoreFunctions<M, WK, WindowState<WV>> storeFunctions = null;
+
+    /**
+     * Public API methods start here
+     */
+
+    /**
+     * Public API method to define the watermark trigger for the window operator
+     *
+     * @param wndTrigger {@link Trigger} function defines the watermark trigger for this {@link SessionWindow}
+     * @return The window operator w/ the defined watermark trigger
+     */
+    @Override
+    public Window<M, WK, WV, WindowOutput<WK, WV>> setTriggers(TriggerBuilder<M, WV> wndTrigger) {
+      this.trigger = wndTrigger.build();
+      return this;
+    }
+
+    private BiFunction<M, Entry<WK, WindowState<WV>>, WindowOutput<WK, WV>> getTransformFunc() {
+      // TODO: actual implementation of the main session window logic, based on the wndKeyFunction, aggregator, and triggers;
+      return null;
+    }
+
+    private WindowFn<M, WK, WindowState<WV>, WindowOutput<WK, WV>> getInternalWindowFn() {
+      return new WindowFn<M, WK, WindowState<WV>, WindowOutput<WK, WV>>() {
+
+        @Override public BiFunction<M, Entry<WK, WindowState<WV>>, WindowOutput<WK, WV>> getTransformFunc() {
+          return SessionWindow.this.getTransformFunc();
+        }
+
+        @Override public Operators.StoreFunctions<M, WK, WindowState<WV>> getStoreFuncs() {
+          return SessionWindow.this.storeFunctions;
+        }
+
+        @Override public Trigger<M, WindowState<WV>> getTrigger() {
+          return SessionWindow.this.trigger;
+        }
+      };
+    }
+  }
+
+  static <M extends Message, WK, WV, WS extends WindowState<WV>, WM extends WindowOutput<WK, WV>> WindowFn<M, WK, WS, WM> getInternalWindowFn(
+      Window<M, WK, WV, WM> window) {
+    if (window instanceof SessionWindow) {
+      SessionWindow<M, WK, WV> sessionWindow = (SessionWindow<M, WK, WV>) window;
+      return (WindowFn<M, WK, WS, WM>) sessionWindow.getInternalWindowFn();
+    }
+    throw new IllegalArgumentException("Input window type not supported.");
+  }
+
+  /**
+   * Public static API methods start here
+   *
+   */
+
+  /**
+   * The public programming interface class for window function
+   *
+   * @param <M>  the type of input {@link Message}
+   * @param <WK>  the type of key to the {@link Window}
+   * @param <WV>  the type of output value in the {@link WindowOutput}
+   * @param <WM>  the type of message in the window output stream
+   */
+  public interface Window<M extends Message, WK, WV, WM extends WindowOutput<WK, WV>> {
```

Comment:
@vjagadish1989 yes. It is logged as SAMZA-1052. I realize that the integration step would remove samza-operator module completely, which creates big conflicts w/ @prateekm 's refactoring effort. Hence, hold up on that.